lib/internal/webstreams/add-abort-signal.js

```diff
@@ -0,0 +1,55 @@
+'use strict';
+
+const { validateAbortSignal } = require('internal/validators');
+const { isWebstream } = require('internal/webstreams/util');
+const { finished } = require('internal/webstreams/finished');
+const { AbortError, codes } = require('internal/errors');
+const { ERR_INVALID_ARG_TYPE } = codes;
+
+function addAbortSignal(signal, stream) {
+  validateInputs(signal, stream);
+
+  const onAbort = () => {
+    stream.abort(new AbortError());
```

Comment:
abort doesn't exist on web streams? Do they?

Reply:
Only on `WritableStream`. On `ReadableStream` it needs to be `cancel()`. 

There is another fundamental problem with this, however... if the `ReadableStream` has a `Reader`, or the `WritableStream` has a `Writer`, their `cancel()` and `abort()` methods cannot be called directly... and there's no public API for knowing if either is locked. If they are, there's no public API for getting the Reader or the Writer and those have to be used for doing the canceling/aborting.

Instead, in order for us to do this, we need to rely on the internal functions and cannot rely on the public API. This also means that these will only work for the core implementations of web streams (userland implementations would not be supported). 

![image](https://user-images.githubusercontent.com/439929/127006209-9854bdc1-1bf8-4298-82f6-4c511e790a85.png)

Reply:
@jasnell do you think this `finished` for webstreams is still a worthwhile development given that this is the case?